types/stylelint/index.d.ts

```diff
@@ -279,20 +279,48 @@ declare namespace stylelint {
 		[key in `expected${string}` | `rejected${string}`]: RuleMessage;
 	};
 
-	type AutofixMessage = { expected: (actual: string, expected: string) => string };
+	type MessageTypes = {
+		autofix: [actual: string, expected: string];
+		pattern: [input: string, pattern: StringOrRegex];
+		max: [object: string, maximum: number];
+	};
+
+	type ExpectedMessage<T> = {
+		expected: (
+			...args: T extends keyof MessageTypes ? MessageTypes[T] : T extends [string] ? T : never
+		) => string;
+	};
+
+	type PropertyMessage<T = never> = {
+		rejected: (
+			...args: [T] extends [never]
```

Comment:
To avoid the distributive nature of TS conditional types we need a non-naked type parameter.
i.e. `T extends never` wouldn't work